lib/trino-collect/src/main/java/io/trino/collect/cache/EvictableCache.java

```diff
@@ -13,184 +13,226 @@
+
+            BiMap<Token<K>, K> tokenToKey = keyToToken.inverse();
+            ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+            for (Map.Entry<Token<K>, V> entry : values.entrySet()) {
+                Token<K> token = entry.getKey();
+
+                // While token.getKey() returns equal key, a caller may expect us to maintain key identity, in case equal keys are still distinguishable.
+                K key = tokenToKey.get(token);
+                checkState(key != null, "No key found for %s in %s when loading %s", token, tokenToKey, keys);
+
+                result.put(key, entry.getValue());
             }
-            catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                throw new RuntimeException("Interrupted", e);
+            return result.buildOrThrow();
+        }
+        catch (Throwable e) {
+            for (Token<K> token : newTokens) {
+                // Failed to load and it was our new token persisted in tokens map.
+                // No cache entry exists for the token (unless concurrent load happened),
+                // so we need to remove it.
+                tokens.remove(token.getKey(), token);
             }
-
-            // Someone else was loading the key, but the load was invalidated.
+            throw e;
         }
     }
 
     @Override
-    public void put(K key, V value)
+    public void refresh(K key)
     {
-        throw new UnsupportedOperationException("The operation is not supported, as in inherently races with cache invalidation. Use get(key, callable) instead.");
+        // The refresh loads a new entry, if it wasn't in the cache yet. Thus, we would create a new Token.
+        // However, dataCache.refresh is asynchronous and may fail, so no cache entry may be created.
+        // In such case we would leak the newly created token.
+        throw new UnsupportedOperationException();
     }
 
     @Override
-    public void invalidate(Object key)
+    public long size()
     {
-        delegate.invalidate(key);
+        return dataCache.size();
     }
 
     @Override
-    public void invalidateAll(Iterable<?> keys)
+    public void cleanUp()
+    {
+        dataCache.cleanUp();
+    }
+
+    @VisibleForTesting
+    int tokensCount()
     {
-        delegate.invalidateAll(keys);
+        return tokens.size();
     }
 
     @Override
-    public void invalidateAll()
+    public void invalidate(Object key)
     {
-        delegate.invalidateAll();
+        @SuppressWarnings("SuspiciousMethodCalls") // Object passed to map as key K
+        Token<K> token = tokens.remove(key);
+        if (token != null) {
+            dataCache.invalidate(token);
+        }
     }
 
     @Override
-    public long size()
+    public void invalidateAll()
     {
-        // Includes entries being computed. Approximate, as allowed per method contract.
-        return delegate.size();
+        dataCache.invalidateAll();
+        tokens.clear();
```

Comment:
Where the differences between `invalidateAll` and `cleanup` come from ?

Assuming many threads I think calling both  `dataCache.invalidateAll` and `tokens.clear` may lead to stale/unaccessible data in dataCache. 
However it should be evicted later so this is not a problem, am i correct?

Reply:
> Assuming many threads I think calling both `dataCache.invalidateAll` and `tokens.clear` may lead to stale/unaccessible data in dataCache.

correct

> However it should be evicted later so this is not a problem, am i correct?

i think so

Reply:
@findepi could you please answer this:

> Where the differences between invalidateAll and cleanup come from ? 

I am trying to understand why are they implemented differently.

Reply:
`cleanup` is internal bookkeeping callback. it's not directly related to invalidation.